kong/runloop/balancer.lua

```diff
@@ -271,12 +278,30 @@ do
           return
         end
 
+        local hostname = tgt.hostname
         local ok, err
-        ok, err = balancer:setAddressStatus(status, tgt.ip, tgt.port, tgt.hostname)
+        ok, err = balancer:setAddressStatus(status, tgt.ip, tgt.port, hostname)
+
+        if not ok then
+          -- Fallback: target was not found in balancer with hostname reported
+          -- by healtchecker, will look for it using the name it was added
+          local upstream = get_upstream_by_id(upstream_id)
+          local history = fetch_target_history(upstream)
```

Comment:
Instead of scanning the target history on every healthchecker event, you can pass `upstream.hostname` as an argument to `attach_healthchecker_to_balancer` and use that inside the `hc_callback`.